classes/local/store/s3/client.php

```diff
@@ -206,17 +253,23 @@ public function test_connection() {
         $connection->success = true;
         $connection->details = '';
 
-        try {
-            $this->client->headBucket(array('Bucket' => $this->bucket));
-        } catch (\Aws\S3\Exception\S3Exception $e) {
+        if (!$this->is_functional()) {
             $connection->success = false;
-            $connection->details = $this->get_exception_details($e);
-        } catch (\GuzzleHttp\Exception\InvalidArgumentException $e) {
-            $connection->success = false;
-            $connection->details = $this->get_exception_details($e);
-        } catch (\Aws\Exception\CredentialsException $e) {
-            $connection->success = false;
-            $connection->details = $this->get_exception_details($e);
+            $connection->details = '';
+        } else {
```

Comment:
I would recommend de-nesting this block, e.g. one if statement with an early return. Everything else should flow as it would in the _ideal case_, matching the function's name/action, which is to test the connection.

I also think the `$connection->details = '';` which be a get_string probably something along the lines of `the s3 client is not functional, possibly because it has not been set up or configured yet`

Reply:
good point about an extra string to explain the failure. But don't think returning earlier in this case would make the code better. However, I have updated the code and moved if block inside try catch.

Reply:
done